# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import numpy as np
import pandas as pd

import pylab as plt
import seaborn as sns

from scipy.interpolate import interp1d

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_percentage_error

import datetime as dt

pd.set_option('display.max_columns', None)


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather.csv` file. 

#### First import it into a data frame called `austin`.

In [2]:
# Your code here
austin = pd.read_csv('../austin_weather.csv')
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

In [4]:
# Your code here
austin.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [5]:
austin._get_numeric_data().describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [6]:
# Your code here
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [7]:
# Your code here
austin.DewPointAvgF.unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [8]:
# Your observation here
# Almost all registers are numeric data, but the dash contained there may be a zero,
# We must first determine if the dash must be changed to a zero or NaN
# then Datatype must be changed to int64


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [9]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [10]:
correct_col_name = ['new_'+x for x in wrong_type_columns]
correct_col_name

['new_DewPointHighF',
 'new_DewPointAvgF',
 'new_DewPointLowF',
 'new_HumidityHighPercent',
 'new_HumidityAvgPercent',
 'new_HumidityLowPercent',
 'new_SeaLevelPressureHighInches',
 'new_SeaLevelPressureAvgInches',
 'new_SeaLevelPressureLowInches',
 'new_VisibilityHighMiles',
 'new_VisibilityAvgMiles',
 'new_VisibilityLowMiles',
 'new_WindHighMPH',
 'new_WindAvgMPH',
 'new_WindGustMPH',
 'new_PrecipitationSumInches']

In [11]:
def conv_to_numeric(series):
    return pd.to_numeric(series, errors='coerce')

In [12]:
# Your code here
austin[correct_col_name] = austin[wrong_type_columns].apply(conv_to_numeric)

In [13]:
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  new_DewPointHighF  \
0                   0.46  Rain , Thunderstorm               67.0   
1                      0                                    43.0   
2                      0                                    31.0   
3                      0                                    36.0   
4                      T                                    44.0   

   new_DewPointAvgF  new_DewPointLowF  new_HumidityHighPercent  \
0              49.0              43.0                     93.0   
1              36.0              28.0                     93.0   
2              27.0              23.0                     76.0   
3              28.0              21.0                     89.0   
4              40.0              36.0                     86.0   

   new_HumidityAvgPercent  new_HumidityLowPercent  \
0                    75.0                    57.0   
1                    68.0                    43.0   
2                    52.0                    27.0   
3                    56.0                    22.0   
4                    71.0                    56.0   

   new_SeaLevelPressureHighInches  new_SeaLevelPressureAvgInches  \
0                           29.86                          29.68   
1                           30.41                          30.13   
2                           30.56                          30.49   
3                           30.56                          30.45   
4                           30.41                          30.33   

   new_SeaLevelPressureLowInches  new_VisibilityHighMiles  \
0                          29.59                     10.0   
1                          29.87                     10.0   
2                          30.41                     10.0   
3                          30.30                     10.0   
4                          30.27                     10.0   

   new_VisibilityAvgMiles  new_VisibilityLowMiles  new_WindHighMPH  \
0    

In [14]:
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 37 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Date                            1319 non-null   object 
 1   TempHighF                       1319 non-null   int64  
 2   TempAvgF                        1319 non-null   int64  
 3   TempLowF                        1319 non-null   int64  
 4   DewPointHighF                   1319 non-null   object 
 5   DewPointAvgF                    1319 non-null   object 
 6   DewPointLowF                    1319 non-null   object 
 7   HumidityHighPercent             1319 non-null   object 
 8   HumidityAvgPercent              1319 non-null   object 
 9   HumidityLowPercent              1319 non-null   object 
 10  SeaLevelPressureHighInches      1319 non-null   object 
 11  SeaLevelPressureAvgInches       1319 non-null   object 
 12  SeaLevelPressureLowInches       13

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [15]:
# Your code here
austin.columns

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'PrecipitationSumInches', 'Events', 'new_DewPointHighF',
       'new_DewPointAvgF', 'new_DewPointLowF', 'new_HumidityHighPercent',
       'new_HumidityAvgPercent', 'new_HumidityLowPercent',
       'new_SeaLevelPressureHighInches', 'new_SeaLevelPressureAvgInches',
       'new_SeaLevelPressureLowInches', 'new_VisibilityHighMiles',
       'new_VisibilityAvgMiles', 'new_VisibilityLowMiles', 'new_WindHighMPH',
       'new_WindAvgMPH', 'new_WindGustMPH', 'new_PrecipitationSumInches'],
      dtype='object')

In [16]:
# austin_new = austin[['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'Events', 'new_DewPointHighF',
#         'new_DewPointAvgF', 'new_DewPointLowF', 'new_HumidityHighPercent',
#         'new_HumidityAvgPercent', 'new_HumidityLowPercent',
#         'new_SeaLevelPressureHighInches', 'new_SeaLevelPressureAvgInches',
#         'new_SeaLevelPressureLowInches', 'new_VisibilityHighMiles',
#         'new_VisibilityAvgMiles', 'new_VisibilityLowMiles', 'new_WindHighMPH',
#         'new_WindAvgMPH', 'new_WindGustMPH', 'new_PrecipitationSumInches']]

In [17]:
austin_new= austin.drop(wrong_type_columns, axis=1)

In [18]:
austin_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Date                            1319 non-null   object 
 1   TempHighF                       1319 non-null   int64  
 2   TempAvgF                        1319 non-null   int64  
 3   TempLowF                        1319 non-null   int64  
 4   Events                          1319 non-null   object 
 5   new_DewPointHighF               1312 non-null   float64
 6   new_DewPointAvgF                1312 non-null   float64
 7   new_DewPointLowF                1312 non-null   float64
 8   new_HumidityHighPercent         1317 non-null   float64
 9   new_HumidityAvgPercent          1317 non-null   float64
 10  new_HumidityLowPercent          1317 non-null   float64
 11  new_SeaLevelPressureHighInches  1316 non-null   float64
 12  new_SeaLevelPressureAvgInches   13

In [19]:
austin_new.isnull().sum()

Date                                0
TempHighF                           0
TempAvgF                            0
TempLowF                            0
Events                              0
new_DewPointHighF                   7
new_DewPointAvgF                    7
new_DewPointLowF                    7
new_HumidityHighPercent             2
new_HumidityAvgPercent              2
new_HumidityLowPercent              2
new_SeaLevelPressureHighInches      3
new_SeaLevelPressureAvgInches       3
new_SeaLevelPressureLowInches       3
new_VisibilityHighMiles            12
new_VisibilityAvgMiles             12
new_VisibilityLowMiles             12
new_WindHighMPH                     2
new_WindAvgMPH                      2
new_WindGustMPH                     4
new_PrecipitationSumInches        124
dtype: int64

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [20]:
# Your code here
austin_new.isnull().any(axis=1)

0       False
1       False
2       False
3       False
4        True
        ...  
1314    False
1315    False
1316    False
1317    False
1318    False
Length: 1319, dtype: bool

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [21]:
# Your code here
austin_new.isnull().sum()

Date                                0
TempHighF                           0
TempAvgF                            0
TempLowF                            0
Events                              0
new_DewPointHighF                   7
new_DewPointAvgF                    7
new_DewPointLowF                    7
new_HumidityHighPercent             2
new_HumidityAvgPercent              2
new_HumidityLowPercent              2
new_SeaLevelPressureHighInches      3
new_SeaLevelPressureAvgInches       3
new_SeaLevelPressureLowInches       3
new_VisibilityHighMiles            12
new_VisibilityAvgMiles             12
new_VisibilityLowMiles             12
new_WindHighMPH                     2
new_WindAvgMPH                      2
new_WindGustMPH                     4
new_PrecipitationSumInches        124
dtype: int64

#### Calculate the ratio of missing rows to total rows

In [22]:
# Your code here
austin_new.isnull().sum() / austin_new.count()

Date                              0.000000
TempHighF                         0.000000
TempAvgF                          0.000000
TempLowF                          0.000000
Events                            0.000000
new_DewPointHighF                 0.005335
new_DewPointAvgF                  0.005335
new_DewPointLowF                  0.005335
new_HumidityHighPercent           0.001519
new_HumidityAvgPercent            0.001519
new_HumidityLowPercent            0.001519
new_SeaLevelPressureHighInches    0.002280
new_SeaLevelPressureAvgInches     0.002280
new_SeaLevelPressureLowInches     0.002280
new_VisibilityHighMiles           0.009181
new_VisibilityAvgMiles            0.009181
new_VisibilityLowMiles            0.009181
new_WindHighMPH                   0.001519
new_WindAvgMPH                    0.001519
new_WindGustMPH                   0.003042
new_PrecipitationSumInches        0.103766
dtype: float64

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [23]:
# Your code here
austin_new.isna().sum()

Date                                0
TempHighF                           0
TempAvgF                            0
TempLowF                            0
Events                              0
new_DewPointHighF                   7
new_DewPointAvgF                    7
new_DewPointLowF                    7
new_HumidityHighPercent             2
new_HumidityAvgPercent              2
new_HumidityLowPercent              2
new_SeaLevelPressureHighInches      3
new_SeaLevelPressureAvgInches       3
new_SeaLevelPressureLowInches       3
new_VisibilityHighMiles            12
new_VisibilityAvgMiles             12
new_VisibilityLowMiles             12
new_WindHighMPH                     2
new_WindAvgMPH                      2
new_WindGustMPH                     4
new_PrecipitationSumInches        124
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [24]:
# Your code here
(124/1319) * 100

9.401061410159212

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [25]:
# Your code here 
austin_new.drop(columns='new_PrecipitationSumInches', inplace=True)

# Print `austin` to confirm the column is indeed removed

austin_new

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      new_DewPointHighF  new_DewPointAvgF  new_DewPointLowF  \
0                  67.0              49.0              43.0   
1                  43.0              36.0              28.0   
2                  31.0              27.0              23.0   
3                  36.0              28.0              21.0   
4                  44.0              40.0              36.0   
...                 ...               ...               ...   
1314               71.0              67.0              61.0   
1315               71.0              64.0              55.0   
1316               72.0              64.0              55.0   
1317               70.0              68.0              63.0   
1318               66.0              61.0              54.0   

      new_HumidityHighPercent  new_HumidityAvgPercent  new_HumidityLowPercent  \
0                        93.0                    75.0                    57.0   
1                        93.0                    68.0                    43.0   
2                        76.0                    52.0                    27.0   
3                        89.0                    56.0                    22.0   
4                        86.0                    71.0                    56.0   
...                       ...                     ...                     ...   
1314                     82.0                    54.0                    25.0   
1315                     87.0                    54.0                    20.0   
1316                     82.0                    51.0                    19.0   
1317                     69.0                    48.0                    27.0   
1318                     64.0                    43.0                    22.0   

      new_SeaLevelPressureHighInches  new_SeaLevelPressureAvgInches  \
0                              29.86                          29.68   
1                              30.41                          30.13   
2                              30.56                          30.49   
3                              30.56                          30.45   
4                              30.41                          30.33   
...                              ...                            ...   
1314                           30.04                          29.97   
1315                           29.97                          29.90   
1316                           29.91                          29.86   
1317                           29.96                          29.91   
1318                           30.04                          29.97   

      new_SeaLevelPressureLowInches  new_VisibilityHighMiles  \
0                             29.59                     10.0   
1                             29.87                     10.0   
2                             30.41                     10.0   
3                             30.30                     10.0   
4                             30.27                     10.0   
...                             ...                      ...   
1314                          29.88                     10.0   
1315                          29.81                 

In [26]:
austin_new.isnull().sum()

Date                               0
TempHighF                          0
TempAvgF                           0
TempLowF                           0
Events                             0
new_DewPointHighF                  7
new_DewPointAvgF                   7
new_DewPointLowF                   7
new_HumidityHighPercent            2
new_HumidityAvgPercent             2
new_HumidityLowPercent             2
new_SeaLevelPressureHighInches     3
new_SeaLevelPressureAvgInches      3
new_SeaLevelPressureLowInches      3
new_VisibilityHighMiles           12
new_VisibilityAvgMiles            12
new_VisibilityLowMiles            12
new_WindHighMPH                    2
new_WindAvgMPH                     2
new_WindGustMPH                    4
dtype: int64

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [27]:
# Your code here
austin_fixed = austin_new.interpolate(method='linear', limit_direction='forward', inplace=False)
austin_fixed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      new_DewPointHighF  new_DewPointAvgF  new_DewPointLowF  \
0                  67.0              49.0              43.0   
1                  43.0              36.0              28.0   
2                  31.0              27.0              23.0   
3                  36.0              28.0              21.0   
4                  44.0              40.0              36.0   
...                 ...               ...               ...   
1314               71.0              67.0              61.0   
1315               71.0              64.0              55.0   
1316               72.0              64.0              55.0   
1317               70.0              68.0              63.0   
1318               66.0              61.0              54.0   

      new_HumidityHighPercent  new_HumidityAvgPercent  new_HumidityLowPercent  \
0                        93.0                    75.0                    57.0   
1                        93.0                    68.0                    43.0   
2                        76.0                    52.0                    27.0   
3                        89.0                    56.0                    22.0   
4                        86.0                    71.0                    56.0   
...                       ...                     ...                     ...   
1314                     82.0                    54.0                    25.0   
1315                     87.0                    54.0                    20.0   
1316                     82.0                    51.0                    19.0   
1317                     69.0                    48.0                    27.0   
1318                     64.0                    43.0                    22.0   

      new_SeaLevelPressureHighInches  new_SeaLevelPressureAvgInches  \
0                              29.86                          29.68   
1                              30.41                          30.13   
2                              30.56                          30.49   
3                              30.56                          30.45   
4                              30.41                          30.33   
...                              ...                            ...   
1314                           30.04                          29.97   
1315                           29.97                          29.90   
1316                           29.91                          29.86   
1317                           29.96                          29.91   
1318                           30.04                          29.97   

      new_SeaLevelPressureLowInches  new_VisibilityHighMiles  \
0                             29.59                     10.0   
1                             29.87                     10.0   
2                             30.41                     10.0   
3                             30.30                     10.0   
4                             30.27                     10.0   
...                             ...                      ...   
1314                          29.88                     10.0   
1315                          29.81                 

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [28]:
# Your code here
austin_fixed.isnull().sum()

Date                              0
TempHighF                         0
TempAvgF                          0
TempLowF                          0
Events                            0
new_DewPointHighF                 0
new_DewPointAvgF                  0
new_DewPointLowF                  0
new_HumidityHighPercent           0
new_HumidityAvgPercent            0
new_HumidityLowPercent            0
new_SeaLevelPressureHighInches    0
new_SeaLevelPressureAvgInches     0
new_SeaLevelPressureLowInches     0
new_VisibilityHighMiles           0
new_VisibilityAvgMiles            0
new_VisibilityLowMiles            0
new_WindHighMPH                   0
new_WindAvgMPH                    0
new_WindGustMPH                   0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [29]:
# Your code here:
austin_fixed.Events.value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [30]:
# Your answer:
austin_fixed.Events.value_counts().nlargest(2)

        903
Rain    192
Name: Events, dtype: int64

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [31]:
austin_fixed.Events.unique()

array(['Rain , Thunderstorm', ' ', 'Rain', 'Fog', 'Rain , Snow',
       'Fog , Rain', 'Thunderstorm', 'Fog , Rain , Thunderstorm',
       'Fog , Thunderstorm'], dtype=object)

In [32]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here

for i in event_list:
    austin_fixed[i] = np.where(austin_fixed['Events'].str.contains(i), 1, 0)

# Print your new dataframe to check whether new columns have been created:

austin_fixed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 24 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Date                            1319 non-null   object 
 1   TempHighF                       1319 non-null   int64  
 2   TempAvgF                        1319 non-null   int64  
 3   TempLowF                        1319 non-null   int64  
 4   Events                          1319 non-null   object 
 5   new_DewPointHighF               1319 non-null   float64
 6   new_DewPointAvgF                1319 non-null   float64
 7   new_DewPointLowF                1319 non-null   float64
 8   new_HumidityHighPercent         1319 non-null   float64
 9   new_HumidityAvgPercent          1319 non-null   float64
 10  new_HumidityLowPercent          1319 non-null   float64
 11  new_SeaLevelPressureHighInches  1319 non-null   float64
 12  new_SeaLevelPressureAvgInches   13

In [33]:
austin_fixed[austin_fixed['Events'].str.contains('Fog')]

Date  TempHighF  TempAvgF  TempLowF                     Events  \
19    2014-01-09         70        62        53                        Fog   
50    2014-02-09         76        59        41                        Fog   
57    2014-02-16         76        68        59                        Fog   
70    2014-03-01         82        68        53                        Fog   
84    2014-03-15         74        68        62                 Fog , Rain   
...          ...        ...       ...       ...                        ...   
1119  2017-01-13         78        67        56                 Fog , Rain   
1123  2017-01-17         59        54        48  Fog , Rain , Thunderstorm   
1127  2017-01-21         81        65        48                        Fog   
1142  2017-02-05         72        62        51                        Fog   
1262  2017-06-05         90        78        66  Fog , Rain , Thunderstorm   

      new_DewPointHighF  new_DewPointAvgF  new_DewPointLowF  \
19                 60.0              55.0              50.0   
50                 53.0              45.0              38.0   
57                 61.0              57.0              53.0   
70                 65.0              57.0              48.0   
84                 64.0              62.0              59.0   
...                 ...               ...               ...   
1119               69.0              63.0              51.0   
1123               55.0              48.0              46.0   
1127               58.0              51.0              33.0   
1142               64.0              59.0              51.0   
1262               70.0              68.0              65.0   

      new_HumidityHighPercent  new_HumidityAvgPercent  new_HumidityLowPercent  \
19                       93.0                    80.0                    66.0   
50                       96.0                    70.0                    44.0   
57                       90.0                    74.0                    57.0   
70                       93.0                    74.0                    54.0   
84                       93.0                    81.0                    68.0   
...                       ...                     ...                     ...   
1119                    100.0                    74.0                    48.0   
1123                    100.0                    86.0                    72.0   
1127                    100.0                    62.0                    23.0   
1142                    100.0                    89.0                    78.0   
1262                    100.0                    72.0                    44.0   

      new_SeaLevelPressureHighInches  new_SeaLevelPressureAvgInches  \
19                             30.18                          30.11   
50                             30.12                          30.05   
57                             30.08                          30.01   
70                             30.02                          29.95   
84                             29.97                          29.86   
...                              ...                            ...   
1119                           30.36                          30.25   
1123                           30.19                          30.11   
1127                           29.63                          29.58   
1142                           30.13                          30.06   
1262                           29.79                          29.74   

      new_SeaLevelPressureLowInches  new_VisibilityHighMiles  \
19                            30.00                     10.0   
50                            29.93                     10.0   
57                            29.94                     10.0   
70                            29.87                     10.0   
84                            29.68                     10.0   
...                             ...                      ...   
1119                          30.15         

In [34]:
# austin_fixed['Fog'] = np.where(austin_fixed['Events'].str.contains('Fog'), 1, 0)
# austin_fixed['Rain'] = np.where(austin_fixed['Events'].str.contains('Rain'), 1, 0)
# austin_fixed['Thunderstorm'] = np.where(austin_fixed['Events'].str.contains('Thunderstorm'), 1, 0)
# austin_fixed['Snow'] = np.where(austin_fixed['Events'].str.contains('Snow'), 1, 0)

In [35]:
austin_fixed

Date  TempHighF  TempAvgF  TempLowF               Events  \
0     2013-12-21         74        60        45  Rain , Thunderstorm   
1     2013-12-22         56        48        39                        
2     2013-12-23         58        45        32                        
3     2013-12-24         61        46        31                        
4     2013-12-25         58        50        41                        
...          ...        ...       ...       ...                  ...   
1314  2017-07-27        103        89        75                        
1315  2017-07-28        105        91        76                        
1316  2017-07-29        107        92        77                        
1317  2017-07-30        106        93        79                        
1318  2017-07-31         99        88        77                        

      new_DewPointHighF  new_DewPointAvgF  new_DewPointLowF  \
0                  67.0              49.0              43.0   
1                  43.0              36.0              28.0   
2                  31.0              27.0              23.0   
3                  36.0              28.0              21.0   
4                  44.0              40.0              36.0   
...                 ...               ...               ...   
1314               71.0              67.0              61.0   
1315               71.0              64.0              55.0   
1316               72.0              64.0              55.0   
1317               70.0              68.0              63.0   
1318               66.0              61.0              54.0   

      new_HumidityHighPercent  new_HumidityAvgPercent  new_HumidityLowPercent  \
0                        93.0                    75.0                    57.0   
1                        93.0                    68.0                    43.0   
2                        76.0                    52.0                    27.0   
3                        89.0                    56.0                    22.0   
4                        86.0                    71.0                    56.0   
...                       ...                     ...                     ...   
1314                     82.0                    54.0                    25.0   
1315                     87.0                    54.0                    20.0   
1316                     82.0                    51.0                    19.0   
1317                     69.0                    48.0                    27.0   
1318                     64.0                    43.0                    22.0   

      new_SeaLevelPressureHighInches  new_SeaLevelPressureAvgInches  \
0                              29.86                          29.68   
1                              30.41                          30.13   
2                              30.56                          30.49   
3                              30.56                          30.45   
4                              30.41                          30.33   
...                              ...                            ...   
1314                           30.04                          29.97   
1315                           29.97                          29.90   
1316                           29.91                          29.86   
1317                           29.96                          29.91   
1318                           30.04                          29.97   

      new_SeaLevelPressureLowInches  new_VisibilityHighMiles  \
0                             29.59                     10.0   
1                             29.87                     10.0   
2                             30.41                     10.0   
3                             30.30                     10.0   
4                             30.27                     10.0   
...                             ...                      ...   
1314                          29.88                     10.0   
1315                          29.81                 

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [36]:
# Your code here

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [37]:
# Your code here
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF               Events  \
0  2013-12-21         74        60        45  Rain , Thunderstorm   
1  2013-12-22         56        48        39                        
2  2013-12-23         58        45        32                        
3  2013-12-24         61        46        31                        
4  2013-12-25         58        50        41                        

   new_DewPointHighF  new_DewPointAvgF  new_DewPointLowF  \
0               67.0              49.0              43.0   
1               43.0              36.0              28.0   
2               31.0              27.0              23.0   
3               36.0              28.0              21.0   
4               44.0              40.0              36.0   

   new_HumidityHighPercent  new_HumidityAvgPercent  new_HumidityLowPercent  \
0                     93.0                    75.0                    57.0   
1                     93.0                    68.0                    43.0   
2                     76.0                    52.0                    27.0   
3                     89.0                    56.0                    22.0   
4                     86.0                    71.0                    56.0   

   new_SeaLevelPressureHighInches  new_SeaLevelPressureAvgInches  \
0                           29.86                          29.68   
1                           30.41                          30.13   
2                           30.56                          30.49   
3                           30.56                          30.45   
4                           30.41                          30.33   

   new_SeaLevelPressureLowInches  new_VisibilityHighMiles  \
0                          29.59                     10.0   
1                          29.87                     10.0   
2                          30.41                     10.0   
3                          30.30                     10.0   
4                          30.27                     10.0   

   new_VisibilityAvgMiles  new_VisibilityLowMiles  new_WindHighMPH  \
0                     7.0                     2.0             20.0   
1                    10.0                     5.0             16.0   
2                    10.0                    10.0              8.0   
3                    10.0                     7.0             12.0   
4                    10.0                     7.0             10.0   

   new_WindAvgMPH  new_WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0             4.0             31.0     0    0     1             1  
1             6.0             25.0     0    0     0             0  
2             3.0             12.0     0    0     0             0  
3             4.0             20.0     0    0     0             0  
4             2.0             16.0     0    0     0             0

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [38]:
# Your code here
austin_fixed.drop(columns='Events', inplace=True)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [39]:
# Your code here
format_date = "%Y-%m-%d"

In [40]:
austin_fixed.Date

0       2013-12-21
1       2013-12-22
2       2013-12-23
3       2013-12-24
4       2013-12-25
           ...    
1314    2017-07-27
1315    2017-07-28
1316    2017-07-29
1317    2017-07-30
1318    2017-07-31
Name: Date, Length: 1319, dtype: object

In [41]:
austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: dt.datetime.strptime(x, format_date).toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [42]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  new_DewPointHighF  new_DewPointAvgF  \
0  735223         74        60        45               67.0              49.0   
1  735224         56        48        39               43.0              36.0   
2  735225         58        45        32               31.0              27.0   
3  735226         61        46        31               36.0              28.0   
4  735227         58        50        41               44.0              40.0   

   new_DewPointLowF  new_HumidityHighPercent  new_HumidityAvgPercent  \
0              43.0                     93.0                    75.0   
1              28.0                     93.0                    68.0   
2              23.0                     76.0                    52.0   
3              21.0                     89.0                    56.0   
4              36.0                     86.0                    71.0   

   new_HumidityLowPercent  new_SeaLevelPressureHighInches  \
0                    57.0                           29.86   
1                    43.0                           30.41   
2                    27.0                           30.56   
3                    22.0                           30.56   
4                    56.0                           30.41   

   new_SeaLevelPressureAvgInches  new_SeaLevelPressureLowInches  \
0                          29.68                          29.59   
1                          30.13                          29.87   
2                          30.49                          30.41   
3                          30.45                          30.30   
4                          30.33                          30.27   

   new_VisibilityHighMiles  new_VisibilityAvgMiles  new_VisibilityLowMiles  \
0                     10.0                     7.0                     2.0   
1                     10.0                    10.0                     5.0   
2                     10.0                    10.0                    10.0   
3                     10.0                    10.0                     7.0   
4                     10.0                    10.0                     7.0   

   new_WindHighMPH  new_WindAvgMPH  new_WindGustMPH  Snow  Fog  Rain  \
0             20.0             4.0             31.0     0    0     1   
1             16.0             6.0             25.0     0    0     0   
2              8.0             3.0             12.0     0    0     0   
3             12.0             4.0             20.0     0    0     0   
4             10.0             2.0             16.0     0    0     0   

   Thunderstorm  
0             1  
1             0  
2             0  
3             0  
4             0

In [56]:
austin_fixed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 24 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   Date                            1319 non-null   datetime64[ns]
 1   TempHighF                       1319 non-null   int64         
 2   TempAvgF                        1319 non-null   int64         
 3   TempLowF                        1319 non-null   int64         
 4   new_DewPointHighF               1319 non-null   float64       
 5   new_DewPointAvgF                1319 non-null   float64       
 6   new_DewPointLowF                1319 non-null   float64       
 7   new_HumidityHighPercent         1319 non-null   float64       
 8   new_HumidityAvgPercent          1319 non-null   float64       
 9   new_HumidityLowPercent          1319 non-null   float64       
 10  new_SeaLevelPressureHighInches  1319 non-null   float64       
 11  new_

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [48]:
# Your code here:
y = pd.DataFrame(austin_new['TempAvgF'])

In [50]:
x = austin_fixed[['Date', 'TempHighF', 'TempLowF', 'new_DewPointHighF',
        'new_DewPointAvgF', 'new_DewPointLowF', 'new_HumidityHighPercent',
        'new_HumidityAvgPercent', 'new_HumidityLowPercent',
        'new_SeaLevelPressureHighInches', 'new_SeaLevelPressureAvgInches',
        'new_SeaLevelPressureLowInches', 'new_VisibilityHighMiles',
        'new_VisibilityAvgMiles', 'new_VisibilityLowMiles', 'new_WindHighMPH',
        'new_WindAvgMPH', 'new_WindGustMPH', 'Snow', 'Fog', 'Rain',
        'Thunderstorm']]
x

Date  TempHighF  TempLowF  new_DewPointHighF  new_DewPointAvgF  \
0     735223         74        45               67.0              49.0   
1     735224         56        39               43.0              36.0   
2     735225         58        32               31.0              27.0   
3     735226         61        31               36.0              28.0   
4     735227         58        41               44.0              40.0   
...      ...        ...       ...                ...               ...   
1314  736537        103        75               71.0              67.0   
1315  736538        105        76               71.0              64.0   
1316  736539        107        77               72.0              64.0   
1317  736540        106        79               70.0              68.0   
1318  736541         99        77               66.0              61.0   

      new_DewPointLowF  new_HumidityHighPercent  new_HumidityAvgPercent  \
0                 43.0                     93.0                    75.0   
1                 28.0                     93.0                    68.0   
2                 23.0                     76.0                    52.0   
3                 21.0                     89.0                    56.0   
4                 36.0                     86.0                    71.0   
...                ...                      ...                     ...   
1314              61.0                     82.0                    54.0   
1315              55.0                     87.0                    54.0   
1316              55.0                     82.0                    51.0   
1317              63.0                     69.0                    48.0   
1318              54.0                     64.0                    43.0   

      new_HumidityLowPercent  new_SeaLevelPressureHighInches  \
0                       57.0                           29.86   
1                       43.0                           30.41   
2                       27.0                           30.56   
3                       22.0                           30.56   
4                       56.0                           30.41   
...                      ...                             ...   
1314                    25.0                           30.04   
1315                    20.0                           29.97   
1316                    19.0                           29.91   
1317                    27.0                           29.96   
1318                    22.0                           30.04   

      new_SeaLevelPressureAvgInches  new_SeaLevelPressureLowInches  \
0                             29.68                          29.59   
1                             30.13                          29.87   
2                             30.49                          30.41   
3                             30.45                          30.30   
4                             30.33                          30.27   
...                             ...                            ...   
1314                          29.97                          29.88   
1315                          29.90                          29.81   
1316                          29.86                          29.79   
1317                          29.91                          29.87   
1318                          29.97                          29.91   

      new_VisibilityHighMiles  new_VisibilityAvgMiles  new_VisibilityLowMiles  \
0                        10.0                     7.0                     2.0   
1                        10.0                    10.0                     5.0   
2                        10.0                    10.0                    10.0   
3                        10.0                    10.0                     7.0   
4                        10.0                    10.0                     7.0   
...                       ...                     ...                     ...   
1314                     10.0            

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [ ]:
#Your code here:


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [51]:
#Your code here:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20)

In [53]:
y_train.shape, y_test.shape

((1055, 1), (264, 1))

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:


# Bonus Challenge 2

As explained in the README, the main purpose of this lab is to show you the typical process of preparing data for machine learning which sometimes takes up 90% of your time. Data cleaning is a valuable skill to learn and you need to be proficient at various techniques including the ones we showed you above as well as others you'll learn in the future. In the real world this skill will help you perform your job successfully and efficiently.

Now that we're done with data praparation, if you want to expeirence what you'll do in the rest 10% of your time, let's make the final leap.

We will use scikit-learn's [*Support Vector Machines*](https://scikit-learn.org/stable/modules/svm.html) to compute the fit of our training data set, the test on our test data set.

#### In the cell below, import `svm` from `sklearn`:

In [ ]:
# Your code here


#### Now, call `svm.SVC.fit()` on `X_train` and `y_train`. Assign the returned value to a variable called `clf` which stands for *classifier*. Then obtain the test score for `X_test` and `y_test` by calling `clf.score()`.

In [ ]:
# Your code here


#### You now see the model fit score of your test data set. If it's extremely low, it means your selected model is not a good fit and you should try other models.

#### In addition to fitting `X_train`, `y_train`, `X_test`, and `y_test`, you can also fit `X_ts_train`, `y_ts_train`, `X_ts_test`, and `y_ts_test` if you completed Bonus Challenge 1.

In [ ]:
# Your code here


#### We hope you have learned a lot of useful stuff in this lab!